src/lib/salad/hll.c

```diff
@@ -0,0 +1,954 @@
+}
+
+/* Estimate the cardinality of the sparsely represented HyperLogLog. */
+static uint64_t
+hll_sparse_estimate(struct hll *hll)
+{
+	/*
+	 * Since the number of pairs is low compared to linear counting
+	 * estimation threshold linear counting is always used.
+	 */
+	hll_sparse_merge_list_with_buffer(hll);
+	size_t n_counters = hll_n_registers(HLL_SPARSE_PRECISION);
+	size_t n_pairs = hll_sparse_pairs_list_size(hll->data);
+	return linear_counting(n_counters, n_counters - n_pairs);
+}
+
+void
+hll_create(struct hll *hll, uint8_t precision,
+	   enum HLL_REPRESENTATION representation)
+{
+	if (representation == HLL_SPARSE &&
+	    precision >= HLL_SPARSE_MIN_PRECISION) {
+		hll_sparse_create(hll, precision);
+	} else {
+		hll_dense_create(hll, precision);
+	}
+}
+
+struct hll *
+hll_new(uint8_t precision, enum HLL_REPRESENTATION representation)
+{
+	assert(hll_is_supported_precision(precision));
+	struct hll *hll = xcalloc(1, sizeof(*hll));
+	hll_create(hll, precision, representation);
+	return hll;
+}
+
+void
+hll_add(struct hll *hll, uint64_t hash)
+{
+	switch (hll->representation) {
+	case HLL_SPARSE:
+		hll_sparse_add(hll, hash);
+		break;
+	case HLL_DENSE:
+		hll_dense_add(hll, hash);
+		break;
+	default:
+		unreachable();
+	}
+}
+
+/* Check if it is possible to merge the dst estimator with the src. */
+static int MAYBE_UNUSED
+hll_is_mergable(const struct hll *dst, const struct hll *src)
+{
+	return dst->dprecision == src->dprecision;
+}
+
+/* Merge a dense estimator with a dense estimator. */
+static void
+hll_merge_dense_with_dense(struct hll *dst, const struct hll *src)
+{
+	const size_t n_registers = hll_n_registers(dst->dprecision);
+	for (size_t idx = 0; idx < n_registers; ++idx) {
+		uint8_t new_rank = hll_dense_register_rank(src, idx);
+		hll_dense_add_to_register(dst, idx, new_rank);
+	}
+}
+
+/* Merge a dense estimator with a sparse estimator. */
+static void
+hll_merge_dense_with_sparse(struct hll *dst, const struct hll *src)
+{
+	pair_t *list_beg = hll_sparse_pairs_list_begin(src->data);
+	pair_t *list_end = hll_sparse_pairs_list_end(src->data);
+	hll_dense_add_pairs(dst, list_beg, list_end);
+	pair_t *buff_beg = hll_sparse_pairs_buffer_begin(src->data);
+	pair_t *buff_end = hll_sparse_pairs_buffer_end(src->data);
+	hll_dense_add_pairs(dst, buff_beg, buff_end);
+}
+
+int
+hll_merge(struct hll *dst, const struct hll *src)
+{
+	if (!hll_is_mergable(dst, src))
+		return -1;
+
+	if (dst->representation == HLL_SPARSE)
+		hll_sparse_to_dense(dst);
+
+	switch (src->representation) {
+	case HLL_SPARSE:
+		hll_merge_dense_with_sparse(dst, src);
+		return 0;
+	case HLL_DENSE:
+		hll_merge_dense_with_dense(dst, src);
+		return 0;
+	default:
+		unreachable();
+	}
+	return 0;
+}
+
+uint64_t
+hll_estimate(struct hll *hll)
+{
+	switch (hll->representation) {
+	case HLL_SPARSE:
+		return hll_sparse_estimate(hll);
+	case HLL_DENSE:
+		return hll_dense_estimate(hll);
+	default:
+		unreachable();
+		return 0;
```

Comment:
Nit: `return 0;` will never be executed, i.e., it's _dead code_.

Reply:
I scared it may be executed in the release build if `unreachable()` expands to `assert(0)`.

Reply:
I think it can be executed if `unreachable()` is expanded to `assert(0)` and `NDEBUG` is defined.

Reply:
Nope, `unreachable()` is expanded to [`__builtin_unreachable()`](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html):
> If control flow reaches the point of the __builtin_unreachable, the program is undefined. It is useful in situations where the compiler cannot deduce the unreachability of the code.

Reply:
After a personal conversation, we decided to add panic after `unreachable()`.